src/Codec/CBOR/Cuddle/Pretty.hs

```diff
@@ -96,32 +117,101 @@ data GroupRender
+
+fillLeft :: Int -> Doc ann -> Doc ann
+fillLeft len doc = spaces (len - renderedLen doc) <> doc
+
+fillRight :: Int -> Doc ann -> Doc ann
+fillRight len doc = doc <> spaces (len - renderedLen doc)
+
+memberKeySep :: MemberKey -> Doc ann
+memberKeySep MKType {} = " => "
+memberKeySep _ = " : "
+
+columnarGroupChoice :: GrpChoice -> Doc ann
+columnarGroupChoice [] = mempty
+columnarGroupChoice groupEntries@(x : xs) =
+  groupIfNoComments (columnar x : fmap (\a -> "," <+> columnar a) xs)
+  where
+    occurrenceIndicatorLength (WithComments ge _) =
+      renderedLen . pretty $ groupEntryOccurrenceIndicator ge
+
+    longestOccurrenceIndicator = maximum $ occurrenceIndicatorLength <$> groupEntries
+
+    memberKeyLength (WithComments (GEType _ (Just mk) _) _) = renderedLen $ pretty mk
+    memberKeyLength _ = 0
+
+    longestMemberKey = maximum $ memberKeyLength <$> groupEntries
+
+    oiDoc oi =
+      fillLeft longestOccurrenceIndicator (pretty oi)
+        <> if longestOccurrenceIndicator > 0 then space else mempty
+
+    memberKeyDoc mmk = fillRight longestMemberKey (pretty mmk) <> fillLeft longestKeySep (memberKeySep mmk)
+
+    memberKeySepLen (WithComments (GEType _ mmk _) _) = renderedLen $ maybe mempty memberKeySep mmk
+    memberKeySepLen _ = 0
+
+    longestKeySep = maximum $ memberKeySepLen <$> groupEntries
+
+    longestLineColumnar =
+      maximum $ renderedLen . columnar' . stripComment <$> groupEntries
+
+    columnar (WithComments ge Nothing) = columnar' ge
+    columnar (WithComments ge (Just cmt)) = fillRight longestLineColumnar (columnar' ge) <+> prettyCommentNoBreak cmt
+
+    columnar' (GEType oi mmk t0) = oiDoc oi <> maybe mempty memberKeyDoc mmk <> pretty t0
+    columnar' (GEGroup oi g) = oiDoc oi <> prettyGroup AsGroup g
+    columnar' (GERef oi n mga) = oiDoc oi <> pretty n <> pretty mga
+
+    mapInit _ [] = []
+    mapInit _ [a] = [a]
+    mapInit f (a : as) = f a : mapInit f as
+
+    groupIfNoComments
+      | any (\(WithComments _ cmt) -> isJust cmt) groupEntries = mconcat . mapInit (<> hardline)
+      | otherwise = vcat
+
+cEncloseSep :: Doc ann -> Doc ann -> Doc ann -> [Doc ann] -> Doc ann
+cEncloseSep lEnc rEnc _ [] = lEnc <> rEnc
+cEncloseSep lEnc rEnc _ [x] =
+  group . align . vcat $
+    [ lEnc <> softspace <> x
+    , rEnc
+    ]
+cEncloseSep lEnc rEnc s (h : tl) =
+  group . align $ vsep ((lEnc <> lSpaces <> group h) : fmap ((s <> space) <>) tl) <> line' <> rEnc
+  where
+    lSpaces = mconcat $ replicate (renderedLen s) softspace
+
 prettyGroup :: GroupRender -> Group -> Doc ann
-prettyGroup gr (Group (NE.toList -> xs)) =
-  align $ encloseSep lp rp " // " (fmap prettyGrpChoice xs)
+prettyGroup gr (Group (toList -> xs)) =
+  cEncloseSep lEnc rEnc "//" $ fmap (group . columnarGroupChoice) xs
   where
-    prettyGrpChoice = encloseSep mempty mempty ", " . fmap pretty
-    (lp, rp) = case gr of
-      AsMap -> ("{", "}")
-      AsArray -> ("[", "]")
-      AsGroup -> ("(", ")")
+    lEnc = case gr of
```

Comment:
Reason for decoupling these rather than using the tuple? I thilnk it makes it harder to parse

Reply:
No good reason, I reverted this change.